MIDAS/lib/TCAL9539/TCAL9539.cpp

```
@@ -0,0 +1,130 @@
+    }
+    uint8_t pin_offset;
+    uint8_t port_idx;
+    if(pin.pin_id >= 0 && pin.pin_id < LOW_BIT_COUNT){
+        port_idx = 0;
+        pin_offset = pin.pin_id;
+    } else if(pin.pin_id >= LOW_BIT_COUNT && pin.pin_id < HIGH_BIT_COUNT){
+        port_idx = 1;
+        pin_offset = pin.pin_id - LOW_BIT_COUNT;
+    } else {
+        Serial.println("Err bad pin index");
+        return {.is_valid = false};
+    }
+    return {.gpio_addres = address, .port_idx=port_idx, .pin_offset=pin_offset, .is_valid=true};
+}
+
+static uint8_t pin_state[3][2] = {{0xff,0xf},{0xff,0xff},{0xff,0xff}};
+static uint8_t pin_config[3][2] = {{0xff,0xf},{0xff,0xff},{0xff,0xff}};
+
+void digitalWrite(GpioAddress pin, int mode){
+    PinAddress addr = decode(pin);
+    if(!addr.is_valid) return;
+
+    uint8_t current_state = pin_state[pin.gpio_id][addr.port_idx];
+    if(mode == HIGH){
+        current_state |= (1 << addr.pin_offset);
+    } else if(mode == LOW){
+        current_state &= ~(1 << addr.pin_offset);
+    } else {
+        Serial.println("Err bad mode");
+        return;
+    }
+
+    WIRE.beginTransmission(addr.gpio_addres);
+    WIRE.write(REG_OUTPUT0 + addr.port_idx);
+    WIRE.write(current_state);
+    pin_state[pin.gpio_id][addr.port_idx] = current_state;
+    if(!WIRE.endTransmission(true)){
+        Serial.println("Err i2c transmission failed");
+        return;
+    }
+
+}
+
+int digitalRead(GpioAddress pin){
+    PinAddress addr = decode(pin);
+    if(!addr.is_valid) return LOW;
+
+    WIRE.beginTransmission(addr.gpio_addres);
+    WIRE.write(REG_INPUT0 + addr.port_idx);
+    if(!WIRE.endTransmission(true)){
+        Serial.println("Err i2c transmission failed");
+        return LOW;
+    }
+    int ct = WIRE.requestFrom(addr.gpio_addres, 1);
+    if(ct != 1){
+        Serial.println("Err i2c transmission failed");
+        return LOW;
+    }
+    uint8_t val = WIRE.read();
+
+    return val & (1 << addr.pin_offset);
+}
+
+void pinMode(GpioAddress pin, int mode){
```

Would it be better to return an error code just in case we need to check for anything on startup?